### ISSUE TITLE
[Feature]: Dynamic reloading config file

### ISSUE THREAD

**Issue by krrishdholakia** [BerriAI/litellm#964]
### The Feature

"I'm searching for a way to perform dynamic model and endpoint discovery, as it's crucial for our ecosystem. It would be beneficial if LiteLLM allowed the dynamic reloading of the config file, for example. That would help a lot."

### Motivation, pitch

user request

**Comment by ghost** [comment]
cc: @chanwit

**Comment by ghost** [comment]
"Our environment uses Kubernetes, and LLM instances are served by the LM-controller. When the LM-controller starts a new LLM instance, we want it to be automatically registered with a proxy. To achieve this, we plan to write a Kubernetes informer that monitors the creation and deletion of LLM instances, captures their endpoint information, and sends it to register the endpoint with the proxy, in this case, LiteLLM.

For LiteLLM to support this dynamic endpoint discovery, it needs to implement endpoint registration and deregistration, such as POST and DELETE at /endpoints, allowing for on-the-fly alias and endpoint registration without restarting the proxy process. From a security standpoint, we would limit this endpoint to local access only, or possibly use a Unix socket."

**Comment by ghost** [comment]
@chanwit just pushed a new `/model/new` endpoint - https://github.com/BerriAI/litellm/commit/72381c3cc29abf600d9b0a499aea0f6d96f62553

<img width="1500" alt="Screenshot 2023-12-01 at 8 36 52 PM" src="https://github.com/BerriAI/litellm/assets/17561003/2623bc54-5d49-40bc-8b86-61a02ffe1fcf">

<img width="1460" alt="Screenshot 2023-12-01 at 8 37 12 PM" src="https://github.com/BerriAI/litellm/assets/17561003/787130ea-6e31-4d61-9f88-e8ca23103ace">

**Comment by ghost** [comment]
I'll update the thread once this is in prod.

**Comment by ghost** [comment]
in prod in v`1.10.0`

**Comment by chanwit** [comment]
Thank you for the ping @krrishdholakia 

Here's my take on the acceptance criteria for this feature.
- [x] There is the POST `/model/new` endpoint in LiteLLM for registering a new LLM instance. The new LLM instance can be used right away without restarting the current LiteLLM server. This AC has been addressed by https://github.com/BerriAI/litellm/commit/72381c3cc29abf600d9b0a499aea0f6d96f62553
- [ ] There is the DELETE `/models/{:id}` endpoint in LiteLLM for removing an existing LLM instance. The instance is removed from the server without restarting.

**Comment by ghost** [comment]
@chanwit `/model/delete` endpoint added - https://github.com/BerriAI/litellm/commit/92b2cbcdc593de1dccc1dd947852460e678b7024

Here's how to use it: 

Step 1: Add a model with an id to your config.yaml 

```yaml
model_list: 
       - model_name: "openai-model"
          litellm_params:
             model: "gpt-3.5-turbo"
          model_info:
             id: "my-model-id"
```

Step 2: Start your proxy 
```bash 
$ litellm /path/to/config.yaml
```

Step 3: Make a **POST** `/model/delete` call to remove this model from your proxy 

```curl 
curl -X POST http://0.0.0.0:8000/model/delete \
-H 'Content-Type: application/json' \
-d '{"id": "my-model-id"}'
```

**Comment by ghost** [comment]
Let me know if this looks good to you 

Will update this ticket once the release is in prod.

**Comment by ghost** [comment]
closing as this is now out

**Comment by ericblue** [comment]
Hi @krrishdholakia Sorry for the delayed response as I'm just testing this out now.

Can you give a valid example of the POST request for a new model?  Following the Swagger example, the following model:

{
  "model_name": "ollama-custom",
  "litellm_params": {
          "model": "ollama/llama2",
          "api_base": "http://localhost:11434",
          "host": "localhost",
          "port": 8000
  },
  "model_info": {
    "id": "my-test-model",
    "mode": "embedding",
    "input_cost_per_token": 0,
    "output_cost_per_token": 0,
    "max_tokens": 4096,
    "base_model": "gpt-3.5-turbo"
  }
}

Results in a server error:

{
  "detail": "Internal Server Error: could not determine a constructor for the tag 'tag:yaml.org,2002:python/object:litellm.proxy._types.ModelInfo'\n  in \"config_1702158172.732068.yaml\", line 7, column 15"
}

It looks like a file is initially created, but model_info is a serialized python object:

model_list:
- litellm_params:
    api_base: http://localhost:11434
    host: localhost
    model: ollama/llama2
    port: 8000
  model_info: !!python/object:litellm.proxy._types.ModelInfo
    __dict__:
      base_model: gpt-3.5-turbo
      id: test
      input_cost_per_token: 0.0
      max_tokens: 4096
      mode: embedding
      output_cost_per_token: 0.0
    __pydantic_extra__:
      additionalProp1: {}
    __pydantic_fields_set__: !!set
      additionalProp1: null
      base_model: null
      id: null
      input_cost_per_token: null
      max_tokens: null
      mode: null
      output_cost_per_token: null
    __pydantic_private__: null
  model_name: ollama-custom

Interestingly enough litellm_params is persisted ok, and both are passed into yaml.dump the same way in the config object

**Comment by ghost** [comment]
that's concerning. Here's how i was testing it - https://github.com/BerriAI/litellm/blob/a9f103995d76b679df0fc76db4d95e9248eeb189/litellm/tests/test_proxy_server.py#L126

I had to skip it on circle ci due to ci/fastapi issues. Thanks for flagging this, i'll have it fixed today + add better ci testing.

**Comment by ghost** [comment]
fix pushed + testing added @ericblue

**Comment by ghost** [comment]
will update this ticket once the fix is in prod.

**Comment by ghost** [comment]
fix live v`1.12.3`+ onwards

**Comment by ec-jt** [comment]
Is there a way to add a model through the API endpoint without it being stored in the database? 

Sending "db_model": false as model info still stores in the database and if I have the env variable STORE_MODEL_IN_DB=False, I can't use the api to register models.
